Add a pip/virtualenv/virtualenvwrapper mission

eeshangarg opened the pull request:
Hello everyone!

Sorry it took this much time for me to submit this, I am preparing for my high school finals and they consume most of my time(and sleep)! I have added installation instructions for Linux, but since I don't have a Mac, I would really appreciate it if someone could help me write them. I already sent an email about this to devel a long time ago.

If any further changes are needed(even if something needs to be rewritten from scratch), I would love to work on it. Thanks a lot for letting me work on this! It really is an honor to learn from and be a part of the amazing OpenHatch community! It feels like home. :-)

Also, one last thing, whenever I try to run "python manage.py test -v2 missions", I can see the tests for all of the other missions finishing successfully, but the tests I wrote for my mission don't even show up. I already had a conversation about this with bugZPDX on IRC, here is a link to the relevant part of the logs <http://irclogs.jackgrigg.com/irc.freenode.net/openhatch/2014-12-30#i_3536877>. I wrote all of the tests and committed them anyway because I was advised that it would be easier for the mentors to look at my code and tell me what it is that I am doing wrong once I have submitted it as a PR.

Thanks to @willingc, @shaunagm, @paulproteus @ehashman for their guidance and encouragement! It means the world to me! :-)

Happy New Year! :-)

Warm regards,
Eeshan Garg

ehashman commented on mysite/missions/templates/missions/main.html:
```diff
@@ -115,6 +115,26 @@ <h3 class="up-top"><img alt='Training Missions' id='training-missions' alt='Trai
+
+    <div class="submodule-container">
+    {% with fully_completed_missions.pipvirtualenv as mission_completed %}
+    {% with "Using pip & virtualenv" as mission_title %}
```
That red highlighting on the ampersand (&) doesn't look too good! Do you need to escape that?

eeshangarg commented:
Hello @ehashman!

Thanks a lot for the review! You're the best! :-)

I tried and escaped the ampersand, but the backslash was visible on the missions page so I removed the ampersand altogether and replaced it with the word "and".

Sorry it took so much time for me to reply, I am not feeling well(I have a fever). But please feel free to give more feedback, I would love to work on it. ;-)

Happy New Year!

Warm regards,
Eeshan Garg

ehashman commented:
Oh, @eeshangarg, sorry, that was just one comment from my first lookthrough. I promise you a real review soon, possibly today :)

willingc commented on mysite/missions/templates/missions/pipvirtualenv/setup.html:
```diff
@@ -0,0 +1,74 @@
+    <p>To install a C compiler on Fedora, at the terminal, type the
+    following command and press Enter:</p>
+    <pre>sudo yum install python-devel</pre>
+
```
@eeshangarg The Python Guide http://docs.python-guide.org/en/latest/starting/install/osx/ has info under the Doing It Right section. The second paragraph and the following note are the important points to make (or just refer to the Python Guide).

willingc commented on mysite/missions/templates/missions/pipvirtualenv/setup.html:
```diff
@@ -0,0 +1,74 @@
+    <q>command not found</q> or something similar.</p>
+
+    <p>If this is the case, the following command should fix it:</p>
+    <pre>source .bash_profile</pre>
```
Probably would be good to `source ~/.bash_profile` or make a reference to execute from their home directory.

willingc commented on mysite/missions/templates/missions/pipvirtualenv/installing_packages.html:
```diff
@@ -0,0 +1,110 @@
+  </pre>
+
+  <p>After running the <kbd>mkvirtualenv</kbd> command above, the name of the
+  virtualenv you created(<kbd>TestEnv</kbd>, in this case) will appear
```
I think it would be good to remove the parens on this line. I think the parens should only be around (TestEnv) in this paragraph when used to illustrate that the parens tell the user that the virtual environment has been activated.

willingc commented on mysite/missions/templates/missions/pipvirtualenv/installing_packages.html:
```diff
@@ -0,0 +1,110 @@
+  <p>After running the <kbd>mkvirtualenv</kbd> command above, the name of the
+  virtualenv you created(<kbd>TestEnv</kbd>, in this case) will appear
+  before your terminal prompt, which will look somewhat similar to
+  <kbd>(TestEnv)C:\Users\username></kbd>(on Windows) or
```
Remove parens around 'on Windows'

willingc commented on mysite/missions/templates/missions/pipvirtualenv/installing_packages.html:
```diff
@@ -0,0 +1,110 @@
+  virtualenv you created(<kbd>TestEnv</kbd>, in this case) will appear
+  before your terminal prompt, which will look somewhat similar to
+  <kbd>(TestEnv)C:\Users\username></kbd>(on Windows) or
+  <kbd>(TestEnv)user@hostname:~ $</kbd>(on Linux/Mac OS X) at this point.
```
As mentioned ^^

willingc commented on mysite/missions/templates/missions/pipvirtualenv/installing_packages.html:
```diff
@@ -0,0 +1,110 @@
+  below. Then, submit!</p>
+
+  <br>
+
```
The form correctly validates the requests version :) The user interface continues to display an empty submission box which is a little confusing on how to proceed to the next step in the mission.

eeshangarg replied:
@willingc: Hello, ma'am!

I have implemented all of the changes you advised me to. However, I am sorry, but I don't quite understand what you mean here about the empty submission box. Once the user submits the correct Requests version, a green label should pop up saying something like "You have successfully completed this part of the mission", and that's how all of the other missions work as well. Does this label not appear for some reason or are you suggesting that the information in the label should be more explicit as to how to move on to the next part of the mission?

It would mean a lot to me if you could please explain a bit further as to what I should do next about this. See you soon! :-)

Warm regards,
Eeshan Garg

willingc replied:
@eeshangarg I will need to double check since I honestly don't remember exactly. I think that the green label popped up about completion but the form box where the pip freeze text should be entered continued to be displayed. Certainly, not a showstopper for this PR if that is the default behavior.

I will review this PR later today. Thank you!

willingc commented on mysite/missions/templates/missions/pipvirtualenv/removing_packages.html:
```diff
@@ -0,0 +1,107 @@
+  be prefixed by the name of a virtualenv. This means that no virtualenv is
+  active at the moment.</p>
+
+  <p>To activate a virtualenv, at the terminal, type <kbd>workon</kbd>,
```
Using virtualenvwrapper, workon will activate a virtualenv. Otherwise with no virtualenvwrapper, it is source TestEnv/bin/activate.

I'm not sure the best way to keep this simple. I should not that I did not run Glyph's script on my system since I didn't wish for it to change my configuration on OS X.

willingc commented:
@eeshangarg Nicely done :)
I have made some comments on some of the files. I didn't review all of the files. But the mission worked well on my Mac (OS X 10.9.5). :cake:

willingc commented:
@eeshangarg Thank you for making the changes and for putting together this useful mission.
Here's what I meant earlier about the form box persisting after a step is completed:
![screen shot 2015-02-01 at 08 11 18](https://cloud.githubusercontent.com/assets/2680980/5992756/4a6c8674-a9eb-11e4-92a0-65e926b61494.png)

I am going to merge this as is since I think it's a solid addition to the training missions. Thanks for your dedication on seeing it through. Nicely done! :sunny: :penguin:

ehashman commented:
Great work, @eeshangarg! :cake:

vipul-sharma20 commented:
Hi, @eeshangarg. I've figured out why `python manage.py test -v2 missions` does not run your tests. You'll have to import your tests in `/mysite/missions/tests.py`. I've also written tests for a mission and was having the same problem. I ran your tests by importing them and it works fine :)
